database/pgsql/layer.go

```diff
@@ -18,300 +18,349 @@ import (
 	"database/sql"
 	"sort"
 
+	"github.com/deckarep/golang-set"
+
 	"github.com/coreos/clair/database"
 	"github.com/coreos/clair/pkg/commonerr"
 )
 
+const (
+	soiLayer = `
+		WITH new_layer AS (
+			INSERT INTO layer (hash)
+			SELECT CAST ($1 AS VARCHAR)
+			WHERE NOT EXISTS (SELECT id FROM layer WHERE hash = $1)
+			RETURNING id
+		)
+		SELECT id FROM new_Layer
+		UNION
+		SELECT id FROM layer WHERE hash = $1`
+
+	findLayerFeatures = `
+		SELECT f.name, f.version, f.version_format, lf.detector_id
+			FROM layer_feature AS lf, feature AS f
+			WHERE lf.feature_id = f.id
+				AND lf.layer_id = $1`
+
+	findLayerNamespaces = `
+		SELECT ns.name, ns.version_format, ln.detector_id
+			FROM layer_namespace AS ln, namespace AS ns
+			WHERE ln.namespace_id = ns.id
+				AND ln.layer_id = $1`
+
+	findLayerID = `SELECT id FROM layer WHERE hash = $1`
+)
+
+// dbLayerNamespace represents the layer_namespace table.
+type dbLayerNamespace struct {
+	layerID     int64
+	namespaceID int64
+	detectorID  int64
+}
+
+// dbLayerFeature represents the layer_feature table
+type dbLayerFeature struct {
+	layerID    int64
+	featureID  int64
+	detectorID int64
+}
+
 func (tx *pgSession) FindLayer(hash string) (database.Layer, bool, error) {
-	var (
-		layer   database.Layer
-		layerID int64
-		ok      bool
-		err     error
-	)
+	layer := database.Layer{Hash: hash}
+	if hash == "" {
+		return layer, false, commonerr.NewBadRequestError("non empty layer hash is expected.")
+	}
 
-	layer.LayerMetadata, layerID, ok, err = tx.findLayer(hash)
+	layerID, ok, err := tx.findLayerID(hash)
+	if err != nil || !ok {
+		return layer, ok, err
+	}
+
+	detectorMap, err := tx.findAllDetectors()
 	if err != nil {
 		return layer, false, err
 	}
 
-	if !ok {
-		return layer, false, nil
+	if layer.By, err = tx.findLayerDetectors(layerID); err != nil {
+		return layer, false, err
+	}
+
+	if layer.Features, err = tx.findLayerFeatures(layerID, detectorMap); err != nil {
+		return layer, false, err
+	}
+
+	if layer.Namespaces, err = tx.findLayerNamespaces(layerID, detectorMap); err != nil {
+		return layer, false, err
 	}
 
-	layer.Features, err = tx.findLayerFeatures(layerID)
-	layer.Namespaces, err = tx.findLayerNamespaces(layerID)
 	return layer, true, nil
 }
 
-func (tx *pgSession) persistLayer(hash string) (int64, error) {
+func sanitizePersistLayerInput(hash string, features []database.LayerFeature, namespaces []database.LayerNamespace, detectedBy []database.Detector) error {
 	if hash == "" {
-		return -1, commonerr.NewBadRequestError("Empty Layer Hash is not allowed")
+		return commonerr.NewBadRequestError("expected non-empty layer hash")
```

Comment:
TODO: in the future ( not this PR ), we should update the way we handle the errors.